crates/ra_hir_ty/src/infer/expr.rs

```diff
@@ -229,17 +232,31 @@ impl<'a> InferenceContext<'a> {
             }
             Expr::Continue => Ty::simple(TypeCtor::Never),
             Expr::Break { expr } => {
-                if let Some(expr) = expr {
-                    // FIXME handle break with value
-                    self.infer_expr(*expr, &Expectation::none());
-                }
-                if let Some(ctxt) = self.breakables.last_mut() {
-                    ctxt.may_break = true;
+                let val_ty = if let Some(expr) = expr {
+                    self.infer_expr(*expr, &Expectation::none())
+                } else {
+                    Ty::unit()
+                };
+
+                let mut has_brkctx = false;
+
+                if self.breakables.last().is_some() {
+                    has_brkctx = true;
                 } else {
                     self.push_diagnostic(InferenceDiagnostic::BreakOutsideOfLoop {
                         expr: tgt_expr,
                     });
                 }
+
+                if has_brkctx {
+                    let last_ty = self.breakables.last().expect("This is a bug").break_ty.clone();
+                    let merged_type = self.coerce_merge_branch(&last_ty, &val_ty);
```

Comment:
instead of having `has_brkctx` and needing to use `expect` multiple times, how about
```rust
let last_type = if let Some(ctxt) = self.breakables.last() {
    ctxt.break_ty.clone()
} else {
    Ty::Unknown
};
let merged_type = self.coerce_merge_branch(&last_type, &val_ty);
if let Some(ctxt) = self.breakables.last_mut() {
    ctxt.may_break = true;
    ctxt.break_ty = merged_type;
} else {
    // push diagnostic ....
}
```

Reply:
Yeah, that loolks much better!